### ISSUE TITLE
Bug in BatchNormalization

### ISSUE THREAD

**Issue by alzhusan** [fchollet/keras#488]
I found the implementation of the `BatchNormalization` is wrong. If the `mode` is 0, "featurewise normalization" is used and two attributes which named  `running_mean` and `running_std` are used to store the mean and STD of the training data. But both of these attributes are implemented as a Theano expression in the `get_output` function. In the test stage, they refer to mean and STD of the input data, not the mean and STD of the training data as we expected. 

To fix it, the `running_mean` and `running_std` should be implemented as a Theano shared value as model's parameters, and should be updated when training. And I noted that the updating mechanism of them is different from the conventional parameters which are updated by their gradient. Therefore, we need to define individual update methods in some especial `Layer`, and the `update` list should be integrated into the conventional one in the training function of the `Model`.

**Comment by EderSantana** [comment]
I saw that too. And for Theano shared variables, you can't do assignment.
I belieive way to have a state that is updated is if we implement BatchNormalization with a scan loop.
Do you have other ideas?

**Comment by alzhusan** [comment]
I change the `BatchNormalization.get_output` into:
```python
def get_output(self, train):
    X = self.get_input(train)

    if self.mode == 0:
        if train:
            m = X.mean(axis=0)
            # manual computation of std to prevent NaNs
            std = T.mean((X-m)**2 + self.epsilon, axis=0) ** 0.5
            X_normed = (X - m) / (std + self.epsilon)

            # I changed the following 5 lines
            self.running_mean = shared_zeros(self.input_shape)
            self.running_std = shared_ones((self.input_shape))

            mean_update = self.momentum * self.running_mean + (1-self.momentum) * m
            std_update = self.momentum * self.running_std + (1-self.momentum) * std
            
            self.updates = [(self.running_mean, mean_update), (self.running_std, std_update)]
        else:
            X_normed = (X - self.running_mean) / (self.running_std + self.epsilon)

    elif self.mode == 1:
        m = X.mean(axis=-1, keepdims=True)
        std = X.std(axis=-1, keepdims=True)
        X_normed = (X - m) / (std + self.epsilon)

    out = self.gamma * X_normed + self.beta
    return out
```

And append the `updates` into the `Model`'s update list, for example in `Graph` model:
```python
def compile(self, optimizer, loss, theano_mode=None):
    ......
    updates = self.optimizer.get_updates(self.params, self.constraints, train_loss)
    # add the especial update rules into the training process
    for m in self.nodes:
        if hasattr(self.nodes[m],'updates'):
            updates.extend(self.nodes[m].updates)
    ......
```

**Comment by fchollet** [comment]
This is fixed; it required some significant framework-wide changes. All layers / containers now support custom "updates" that they define in a `init_update` method run after the layer has been connected.
